### ISSUE TITLE
It doesn't seem like it's using GPU

### ISSUE THREAD

**Issue by ug0x01** [EMI-Group/tensorneat#16]
Hello, I just ran one of the examples and when I check my task manager, it doesn't seem like it's using GPU at all

![Image](https://github.com/user-attachments/assets/764451ec-8cc1-4c4d-b45d-5560b2b94819)

I'm using NVIDIA 1650 and cuda is avilable in my system

Example that I used:

```Python
# Import necessary modules
from tensorneat.pipeline import Pipeline
from tensorneat.algorithm.neat import NEAT
from tensorneat.genome import DefaultGenome, BiasNode
from tensorneat.problem.rl import BraxEnv
from tensorneat.common import ACT, AGG

# Define the pipeline
pipeline = Pipeline(
    algorithm=NEAT(
        pop_size=1000,
        species_size=20,
        survival_threshold=0.1,
        compatibility_threshold=1.0,
        genome=DefaultGenome(
            num_inputs=11,
            num_outputs=3,
            init_hidden_layers=(),
            node_gene=BiasNode(
                activation_options=ACT.tanh,
                aggregation_options=AGG.sum,
            ),
            output_transform=ACT.tanh,
        ),
    ),
    problem=BraxEnv(
        env_name="hopper",
        max_step=1000,
    ),
    seed=42,
    generation_limit=100,
    fitness_target=5000,
)

# Initialize state
state = pipeline.setup()

# Run until termination
state, best = pipeline.auto_run(state)
```

**Comment by ug0x01** [comment]
Hello @WLS2002, Thanks for the reply and also I would like to ask one more general question about NEAT topology itself, at gradient based backpropagation algorithms, bias makes sens to be used because while training if a neuron even once becomes 0, it won't be able to recover and that neuron couldn't be trained but for a NEAT algorithm, it chooses/replaces weights randomly after a mutate and since there is no backpropagation step, I don't understand what's the point of using bias in a NEAT algorithm. Thanks

**Comment by ug0x01** [comment]
Since I couldn't find that much advanced resource about this topology, I just wanted to ask a few question about it to you :) 

1. I wonder why a "base network" isn't being used instead of starting with 0 node network. F.E if it's a complex problem, we'll need a deep NN stracture with big node numbers and untill getting to that point, why aren't we starting from a schema of starting stracture like 3 layers each one includes 64 neurons etc.?
2. I don't understand how species are being used. I mean we are basicly choosing the fittest network and modifying the network with some kind of mutation algorithms and it gets better and better each generation but I don't understand how species are defined. What exactly specie is and how we create them?
3. Finally I wonder what should we do if we face with overfitting (other then adding random noice and I also assume that we are using already pretty big datasets)?

Also I'd be happy if you can give some resources to learn the topology with details. Thank you so much from now

**Comment by WLS2002** [comment]
Hi,  

Regarding the bias question, your understanding is correct. In NEAT, bias is not a necessary part of evolution. However, the presence of bias increases the flexibility of network computations, allowing for better adaptability to different tasks. It helps stabilize the outputs and improves network expressiveness, especially when dealing with non-linear transformations.  

Regarding the "base network" question, your intuition is also correct. In practical applications of NEAT, we don’t always start with the simplest network. Instead, we may use an initial network with multiple hidden layers and either grow or prune (check this interesting [demo](https://github.com/EMI-Group/interpretable-control-competition-walker-submit)) it as evolution progresses. 

For the role of species and how they are formed, you can find more details in the [NEAT origin paper](https://ieeexplore.ieee.org/abstract/document/6790655). It explains how species are used to maintain diversity and prevent premature convergence by grouping similar genomes together based on genetic distance.  

Lastly, regarding overfitting, I find it difficult to provide a definitive answer. I primarily use NEAT for reinforcement learning tasks, and I haven’t personally encountered overfitting issues in my work.  

If you're interested in learning more about neuroevolution, you might find the paper [Designing Neural Networks Through Neuroevolution](https://www.nature.com/articles/s42256-018-0006-z) uesful. This paper provides a comprehensive overview of how neuroevolution techniques, including NEAT, are used to design neural networks. It discusses different evolutionary strategies, network representations, and how evolution-based approaches compare with traditional deep learning methods.  

Hope this helps! Let me know if you have more questions.

**Comment by sopotc** [comment]
Was passing by and thought I'd add my thoughts hoping they'll be useful.

> bias makes sens to be used because while training if a neuron even once becomes 0, it won't be able to recover and that neuron couldn't be trained

This is not why bias is used in gradient-based algorithms. Think of bias as moving the activation 'bar' of each neuron. Without bias you can't train a one-weight model to learn the function `y=w1*x +3`, let alone more complex functions. No bias means that the function learned has to pass through the origin which is severely limiting.


On the question of why do we start so small - that is a key feature of NEAT. There is a very interesting [paper](https://arxiv.org/abs/1107.0037) from the same authors of NEAT that analyses this empirically. The short version is that starting with a large network is suboptimal, and it is likely that NEAT will find a better solution if it is allowed to complexify from scratch, rather than bias the search with an already existing model. Additionally NEAT _only adds_ complexity, and does not remove it and this is by design (also mentioned in the paper).

**Comment by ug0x01** [comment]
Hello @sopotc, thank you for your explanation!

>  Without bias you can't train a one-weight model to learn the function `y=w1*x +3`

Okay now I understand the point of bias. 

> The short version is that starting with a large network is suboptimal, and it is likely that NEAT will find a better solution if it is allowed to complexify from scratch, rather than bias the search with an already existing model

The complex problems are more likely getting solved by deep networks, I want to ask how we end up with a deep network if we do not start from a base network (I'm sorry if there is an "add layer" functionality that i didn't see). How that function should be implemented?

Thanks!
